WordPress/gutenberg
# Selectors API: Fix for global styles hook, style variations, and duotone

Related:
- https://github.com/WordPress/gutenberg/pull/46496
- https://github.com/WordPress/gutenberg/issues/45194
- https://github.com/WordPress/wordpress-develop/pull/3942

Made obsolete by this PR:
- https://github.com/WordPress/gutenberg/pull/49325
- https://github.com/WordPress/gutenberg/pull/49360

## What?

https://github.com/WordPress/gutenberg/pull/46496 landed in a broken state. It missed addressing the following:
- use of the selectors API in the JS global styles hook: `useGlobalStylesOutput`
- updating Global Styles created style variations and their selectors
- updating the image block and duotone support to use the Selectors API

This PR aims to fix the broken functionality and add the missing pieces that should have been in #46496.

In addition, [some concerns](https://github.com/WordPress/gutenberg/pull/46496#discussion_r1149645412) were raised around the introduction of the editor-only selectors and whether the need warrants having to support that part of the API moving forward. In an effort to unblock the general Selectors API, the editor selectors functionality has been removed in this PR. It can be added in a follow-up and discussed separately.

## Why?

The [Selectors API](https://github.com/WordPress/gutenberg/pull/46496) introduced a few regressions, that need fixing. This PR is one pathway to that. The alternative to this PR is to revert #46496 completely.

## How?

1. Removed the editor-only selectors functionality from what had already landed for the Selectors API
2. Moved the Duotone support from the `filters` key to `filter`
3. Refactored the functions helping generate style declarations so they could be reused for style variations
4. Updated the PHP theme.json class processing of style variations to handle the Selectors API shape.
5. Updated `useGlobalStylesOutput` to use the Selectors API
6. Finished off updating Duotone support to use the Selectors API

**This PR might be easiest to review commit-by-commit as they should match each item from the list above.**

## Next Steps

- [x] Decide on whether to proceed here or revert #46496 
- [ ] (Follow-up) Improve the application of borders on inline image cropper.

## Testing Instructions

1. Run unit tests
    - `npm run test:unit packages/block-editor/src/components/global-styles/test/use-global-styles-output.js`
    - `npm run test:unit packages/blocks/src/api/test/registration`
    - `npm run test:unit:php -- --filter WP_Get_Block_CSS_Selector_Test`
    - `npm run test:unit:php -- --filter WP_Theme_JSON_Gutenberg_Test`
    - `npm run test:unit:php -- --filter WP_Duotone_Gutenberg_Test` 
3. Confirm theme.json styling of image block works as expected
4. Confirm Global Styles for the image block are reflected in the site editor, block editor, and frontend
5. Check Global Styles for other blocks continue to work correctly in both editors and the frontend
6. Create some Block Style variations via Global Styles for multiple block types with one being the Image block
7. Confirm that the style variations you created update correctly in real-time in the site editor and get applied as expected in the block editor and frontend
8. Test duotone filters work on Images in the site editor
9. Switch to the block editor and confirm individual block duotone filters work and display on the frontend

**Note: There's an existing issue with Global Styles duotone filters not applying in the block editor ([fix underway](https://github.com/WordPress/gutenberg/pull/49239)).**
